File: YASiOS9AudioUnitOutputSample/YASiOS9AudioUnitOutputSample/AudioUnitGeneratorSample.m
//
//  AudioUnitGeneratorSample.m
//  YASiOS9AudioUnitOutputSample
//
//  Created by Yuki Yasoshima on 2015/09/20.
//  Copyright © 2015年 Objective-Audio. All rights reserved.
//

#import "AudioUnitGeneratorSample.h"
#import <Accelerate/Accelerate.h>

NS_ASSUME_NONNULL_BEGIN

@interface AudioUnitGeneratorSampleKernel : NSObject
@property (nullable) AVAudioPCMBuffer *outputBuffer;
@property (copy) void (^renderBlock)(AVAudioPCMBuffer *buffer);
@end

@implementation AudioUnitGeneratorSampleKernel

- (void)setupBufferWithFormat:(AVAudioFormat *)format frames:(NSUInteger)frames
{
    self.outputBuffer = [[AVAudioPCMBuffer alloc] initWithPCMFormat:format frameCapacity:frames];
}

- (void)disposeBuffer
{
    self.outputBuffer = nil;
}

@end

@interface AudioUnitGeneratorSample ()

@property AudioUnitGeneratorSampleKernel *kernel;
@property (copy) AUInternalRenderBlock internalBlock;

@property AUAudioUnitBusArray *inputBusArray;
@property AUAudioUnitBusArray *outputBusArray;

@end

@implementation AudioUnitGeneratorSample

+ (AudioComponentDescription)audioComponentDescription
{
    AudioComponentDescription acd = {
        // Generatorはインプットの接続なしでレンダーされる。Effectはインプットに接続しないとレンダーされない
        .componentType = kAudioUnitType_Generator,
        .componentSubType = 'smpl',
        .componentManufacturer = 'Demo',
        .componentFlags = 0,
        .componentFlagsMask = 0,
    };
    return acd;
}

+ (void)initialize
{
    // AudioUnitGeneratorSampleをAudioUnitシステムに登録する
    // サンプル用なのでExtensionではやってはいけない

    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [self registerSubclass:[AudioUnitGeneratorSample class]
            asComponentDescription:[self audioComponentDescription]
                              name:@"SampleAudioUnitName"
                           version:UINT32_MAX];
    });
}

- (nullable instancetype)initWithComponentDescription:(AudioComponentDescription)componentDescription
                                              options:(AudioComponentInstantiationOptions)options
                                                error:(NSError **)outError
{
    self = [super initWithComponentDescription:componentDescription options:options error:outError];
    if (self) {
        // バックグラウンドでのオーディオ処理用のオブジェクトを生成
        AudioUnitGeneratorSampleKernel *kernel = [[AudioUnitGeneratorSampleKernel alloc] init];
        self.kernel = kernel;

        self.internalBlock =
            ^AUAudioUnitStatus(AudioUnitRenderActionFlags *actionFlags, const AudioTimeStamp *timeStamp,
                               AVAudioFrameCount frameCount, NSInteger outputBusNumber, AudioBufferList *outputData,
                               const AURenderEvent *realtimeEventListHead, AURenderPullInputBlock pullInputBlock) {
                // このブロックの中はオーディオのレンダースレッドから呼ばれる

                AVAudioPCMBuffer *buffer = kernel.outputBuffer;
                buffer.frameLength = frameCount;

                void (^renderBlock)(AVAudioPCMBuffer *buffer) = kernel.renderBlock;

                if (renderBlock) {
                    renderBlock(buffer);
                }

                for (UInt32 i = 0; i < outputData->mNumberBuffers; ++i) {
                    void *outData = outputData->mBuffers[i].mData;
                    void *inData = buffer.mutableAudioBufferList->mBuffers[i].mData;
                    if (outData == NULL) {
                        outputData->mBuffers[i].mData = inData;
                    } else if (outData != inData) {
                        memcpy(outData, inData, outputData->mBuffers[i].mDataByteSize);
                    }
                }

                return noErr;
            };

        // バスを作る（アウトプットがひとつ）
        AVAudioFormat *format = [[AVAudioFormat alloc] initStandardFormatWithSampleRate:44100.0 channels:2];
        AUAudioUnitBus *outputBus = [[AUAudioUnitBus alloc] initWithFormat:format error:nil];
        self.outputBusArray =
            [[AUAudioUnitBusArray alloc] initWithAudioUnit:self busType:AUAudioUnitBusTypeOutput busses:@[outputBus]];
    }
    return self;
}

- (AUInternalRenderBlock)internalRenderBlock
{
    NSLog(@"call internalRenderBlock");

    return self.internalBlock;
}

- (void)setRenderBlock:(void (^)(AVAudioPCMBuffer *))renderBlock
{
    NSLog(@"call setRenderBlock");

    self.kernel.renderBlock = renderBlock;
}

- (AUAudioUnitBusArray *)outputBusses
{
    NSLog(@"call outputBusses");

    return self.outputBusArray;
}

- (BOOL)shouldChangeToFormat:(AVAudioFormat *)format forBus:(AUAudioUnitBus *)bus
{
    NSLog(@"call shouldChangeToFormat:forBus:");

    return YES;
}

- (BOOL)allocateRenderResourcesAndReturnError:(NSError *_Nullable __autoreleasing *)outError
{
    NSLog(@"call allocateRenderResourcesAndReturnError");

    BOOL result = [super allocateRenderResourcesAndReturnError:outError];

    if (result) {
        AUAudioUnitBus *bus = self.outputBusArray[0];
        [self.kernel setupBufferWithFormat:bus.format frames:self.maximumFramesToRender];
    }

    return result;
}

- (void)deallocateRenderResources
{
    NSLog(@"call deallocateRenderResources");

    [self.kernel disposeBuffer];

    [super deallocateRenderResources];
}

@end

NS_ASSUME_NONNULL_END

File: iOS9AudioUnitOutputSample/iOS9AudioUnitOutputSample/AudioUnitGeneratorSample.swift
//
//  AudioUnitGeneratorSample.swift
//  iOS9AudioUnitOutputSample
//
//  Created by 八十嶋祐樹 on 2015/11/23.
//  Copyright © 2015年 Yuki Yasoshima. All rights reserved.
//

import AVFoundation

class AudioUnitGeneratorSample: AUAudioUnit {
    
    // MARK: - Private
    
    private let _kernel: AudioUnitSampleKernel = AudioUnitSampleKernel()
    private var _outputBusArray: AUAudioUnitBusArray?
    private var _internalRenderBlock: AUInternalRenderBlock?
    
    // MARK: - Global
    
    static let audioComponentDescription = AudioComponentDescription(
        componentType: kAudioUnitType_Generator,
        componentSubType: hfsTypeCode("gnsp"),
        componentManufacturer: hfsTypeCode("Demo"),
        componentFlags: 0,
        componentFlagsMask: 0
    );
    
    override static func initialize() {
        struct Static {
            static var token: dispatch_once_t = 0
        }
        
        dispatch_once(&Static.token) {
            AUAudioUnit.registerSubclass(
                self,
                asComponentDescription: AudioUnitGeneratorSample.audioComponentDescription,
                name: "AudioUnitGeneratorSample",
                version: UINT32_MAX
            )
        }
    }
    
    // MARK: - Override
    
    override init(componentDescription: AudioComponentDescription, options: AudioComponentInstantiationOptions) throws {
        let kernel = self._kernel
        
        self._internalRenderBlock = { (actionFlags, timeStamp, frameCount, outputBusNumber, outputData, renderEvent, pullInputBlock) in
            
            guard let buffer = kernel.buffer else {
                return noErr
            }
            
            buffer.frameLength = frameCount
            
            if let renderBlock = kernel.renderBlock {
                renderBlock(buffer: buffer)
            }
            
            let out_abl = UnsafeMutableAudioBufferListPointer(outputData)
            let in_abl = UnsafeMutableAudioBufferListPointer(buffer.mutableAudioBufferList)
            
            for i in 0..<out_abl.count {
                let out_data = out_abl[i].mData
                let in_data = in_abl[i].mData
                
                if out_data == nil {
                    out_abl[i].mData = in_data
                } else if out_data != in_data {
                    memcpy(out_data, in_data, Int(out_abl[i].mDataByteSize))
                }
            }
            
            return noErr
        }
        
        do {
            try super.init(componentDescription: componentDescription, options: options)
            
            let format = AVAudioFormat(standardFormatWithSampleRate: 44100.0, channels: 2)
            let bus = try AUAudioUnitBus(format: format)
            self._outputBusArray = AUAudioUnitBusArray(audioUnit: self, busType: AUAudioUnitBusType.Output, busses: [bus])
        } catch {
            throw error
        }
    }
    
    override var outputBusses : AUAudioUnitBusArray {
        get {
            return self._outputBusArray!
        }
    }
    
    override var internalRenderBlock: AUInternalRenderBlock {
        get {
            return self._internalRenderBlock!
        }
    }
    
    override func shouldChangeToFormat(format: AVAudioFormat, forBus bus: AUAudioUnitBus) -> Bool {
        return true
    }
    
    override func allocateRenderResources() throws {
        do {
            try super.allocateRenderResources()
        } catch {
            throw error
        }
        
        let bus = self.outputBusses[0]
        _kernel.buffer = AVAudioPCMBuffer(PCMFormat: bus.format, frameCapacity: self.maximumFramesToRender)
    }
    
    override func deallocateRenderResources() {
        _kernel.buffer = nil
    }
    
    // MARK: - Accessor
    
    var kernelRenderBlock: KernelRenderBlock? {
        get {
            return _kernel.renderBlock
        }
        set {
            _kernel.renderBlock = newValue
        }
    }
}
